torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -891,7 +891,7 @@ def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
       # This can be used to debug FSDP parameter memory consumption.
       outputs = self._dummy_forward(*args, **kwargs)
 
-    if self.reshard_after_forward:
+    if self.reshard_after_forward or not self.training:
```

Comment:
A concern regarding `self.training` change here and below (and the change in test/test_train_mp_imagenet_fsdp.py) -- I feel we should still allow users to call backward on a model in eval mode, but with the change here, such cases would crash.

I think there could be legitimate use cases where the user want to do back-propagation in such cases (e.g. one of them is [domain adaptation](https://stackoverflow.com/questions/57323023/pytorch-loss-backward-and-optimizer-step-in-eval-mode-with-batch-norm-laye) and there are also cases mentioned on PyTorch [forum](https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615)).

Reply:
This is an interesting use cases that I didn't think of. If that's the case, I didn't have any better solution to address @alanwaketan 's TODO, will revert it back.